Disconnect from outbound peers with bad headers chains

sdaftuar opened the pull request:
The first commit will disconnect an outbound peer that serves us a headers chain with insufficient work while we're in IBD.

The second commit introduces a way to disconnect outbound peers whose chains fall out of sync with ours:

For a given outbound peer, we check whether their best known block (which is known from the blocks they announce to us) has at least as much work as our tip.  If it doesn't, we set a 20 minute timeout, and if we still haven't heard about a block with as much work as our tip had when we set the timeout, then we send a single getheaders message, and wait 2 more minutes.  If after two minutes their best known block has insufficient work, we disconnect that peer.

We protect 4 of our outbound peers (who provide some "good" headers chains, ie a chain with at least as much work as our tip at some point) from being subject to this logic, to prevent excessive network topology changes as a result of this algorithm, while still ensuring that we have a reasonable number of nodes not known to be on bogus chains.

We also don't require our peers to be on the same chain as us, to prevent accidental partitioning of the network in the event of a chain split.  Note that if our peers are ever on a more work chain than our tip, then we will download and validate it, and then either reorg to it, or learn of a consensus incompatibility with that peer and disconnect.  This PR is designed to protect against peers that are on a less work chain which we may never try to download and validate.

sdaftuar committed:
Disconnecting from bad outbound peers in IBD

When in IBD, we'd like to use all our outbound peers to help us
sync the chain.  Disconnect any outbound peers whose headers have
insufficient work.

sdaftuar committed:
Permit disconnection of outbound peers on bad/slow chains

Currently we have no rotation of outbound peers.  If an outbound peer
stops serving us blocks, or is on a consensus-incompatible chain with
less work than our tip (but otherwise valid headers), then we will never
disconnect that peer, even though that peer is using one of our 8
outbound connection slots.  Because we rely on our outbound peers to
find an honest node in order to reach consensus, allowing an
incompatible peer to occupy one of those slots is undesirable,
particularly if it is possible for all such slots to be occupied by such
peers.

Protect against this by always checking to see if a peer's best known
block has less work than our tip, and if so, set a 20 minute timeout --
if the peer is still not known to have caught up to a chain with as much
work as ours after 20 minutes, then send a single getheaders message,
wait 2 more minutes, and if a better header hasn't been received by then,
disconnect that peer.

Note:

- we do not require that our peer sync to the same tip as ours, just an
equal or greater work tip.  (Doing otherwise would risk partitioning the
network in the event of a chain split, and is also unnecessary.)

- we pick 4 of our outbound peers and do not subject them to this logic,
to be more conservative. We don't wish to permit temporary network
issues (or an attacker) to excessively disrupt network topology.

sipa commented:
I've spent some time verifying the logic, and it looks very good.

I briefly misunderstood it to mean that if a peer makes _progress_ during the 20 minute window, it would be extended. That would have been a problem (an incompatible chain that has less work but does grow continuously), but it's not what happens (it requires reaching the same amount of work as we had at the beginning of the window).

A test for the second commit would be great.

gmaxwell commented on src/net_processing.cpp:
```diff
@@ -2363,6 +2363,23 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+                // whitelisted).
+                if (!(pfrom->fInbound || pfrom->fWhitelisted || pfrom->fAddnode)) {
+                    pfrom->fDisconnect = true;
+                }
```
I think this needs to be backed off a bit:  Consider-- imagine we start and are fully synced up, but we are still in IsInitialBlockDownload because the network has been slow and so the tip's timestamp is a ways back.  Then we have a peer we get headers from who is a single block behind us.   This would disconnect them, and I think that is probably undesirable.

It's probably a good idea to review all code as if IsInitialBlockDownload() were just return true; and if there would be misbehavior in that case, then the usage is perhaps suboptimal.
An obvious conservative thing to do would be to just use the nMinimumChainWork ... though perhaps too conservative.  It might also be reasonable to use a GetBlockProofEquivalentTime like test, or just the work several blocks back from chainactive tip.   I think if it was conservative enough it could apply at all times, not just IsIninitialBlockDownload().  

Maybe the existence of the second commit means that it would be okay for this one to be less aggressive?

sdaftuar replied:
I guess with the second commit, we will eventually disconnect some peers if they stay on chains with less work than our tip.  So I think we could just leave this as nMinimumChainWork, to protect against the case that all our peers only have incomplete/bogus chains (which wouldn't be prevented by the second commit).

promag replied:
Does it make sense to avoid the push above:
https://github.com/bitcoin/bitcoin/blob/e9d58023aef97d954b7a75da806ee4f22c9760c3/src/net_processing.cpp#L2332
now that the node will disconnect?

sdaftuar commented:
Thanks for the review.  I've pushed up a couple quick fixes for the nits.  Will work on a test, as well as trying to implement a suggestion @gmaxwell gave me offline for improving the behavior in the situation where none of our peers are giving us any block announcements.

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -3247,6 +3290,55 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+                    pto->fDisconnect = true;
+                } else {
+                    const CBlockIndex *pindexStart = state.header_with_required_work;
+                    assert (pindexStart != nullptr);
```
Ugh, just for a print?

TheBlueMatt reviewed (commented):
Looks pretty good. This is begging for a clearer node-sync-state model, but its fine for 0.15.0.2. Does want a test.

sdaftuar committed:
Add unit test for outbound peer eviction

sdaftuar reviewed (commented):
Thanks for the review -- pushed up some fixup commits, which I'll squash.

I switched to using GetTime() everywhere, and was then able to write up a unit test.  This is currently next-to-impossible to test in the regtest environment, I believe, since we don't support non-manual connections in regtest.

sdaftuar commented:
Squashed (https://github.com/sdaftuar/bitcoin/commits/11490.1 -> 032dc531b911a0c0ea570aa74531508f54a36073)

jonasschnelli commented:
Travis found:
```
test/DoS_tests.cpp(80): error in "outbound_slow_chain_eviction": check dummyNode1.fDisconnect == true failed
```

sdaftuar commented:
Unit test issue should be fixed now.

promag commented on src/net_processing.cpp:
```diff
@@ -534,6 +549,8 @@ void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTim
     nPreferredDownload -= state->fPreferredDownload;
     nPeersWithValidatedDownloads -= (state->nBlocksInFlightValidHeaders != 0);
     assert(nPeersWithValidatedDownloads >= 0);
+    g_outbound_peers_with_protect_from_disconnect -= state->m_protect_from_disconnect;
```
`int -= bool`?

promag replied:
More verbose but IMO more clear:
```cpp
if (state->m_protect_from_disconnect) {
    assert(g_outbound_peers_with_protect_from_disconnect > 0);
    --g_outbound_peers_with_protect_from_disconnect;
}
```

sdaftuar replied:
I was following the pattern already in use in the lines above (`nPreferredDownload -= state->fPreferredDownload`).  But I can change to something like what you suggest if this is frowned upon for some reason.

promag commented on src/net_processing.cpp:
```diff
@@ -124,6 +124,9 @@ namespace {
     int nPeersWithValidatedDownloads = 0;
 
+    /** Number of outbound peers with m_protect_from_disconnect. */
+    int g_outbound_peers_with_protect_from_disconnect = 0;
```
Remove since this is easily calculated by counting nodes with `m_protect_from_disconnect`?

sdaftuar replied:
By looping over all the CNodeState objects when this value is needed?  I don't think that makes sense.  Note that tracking the value here is analogous to what we do with `nPeersWithValidatedDownloads` and `nPreferredDownload`.

promag commented on src/net_processing.h:
```diff
@@ -21,6 +21,12 @@ static const unsigned int DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN = 100;
+/** Protect at least this many outbound peers from disconnection due to slow/
+ * behind headers chain.
+ */
+static constexpr int32_t MAX_OUTBOUND_PEERS_TO_PROTECT = 4;
```
`_FROM_DISCONNECT`?

promag commented on src/net_processing.cpp:
```diff
@@ -2383,6 +2383,21 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+        // If we're in IBD, we want outbound peers that will serve us a useful
+        // chain.  Disconnect peers that are behind us or on chains with
+        // insufficient work.
+        if (IsInitialBlockDownload() && nCount != MAX_HEADERS_RESULTS) {
```
Swap conditions as the second is faster?

sdaftuar replied:
I don't think this really matters?

TheBlueMatt replied:
> if we receive fewer than MAX_HEADERS_RESULTS, then we have no reason to request more headers in order to update pindexBestKnownBlock. Note that in the logic below, we are using pindexBestKnownBlock, not pindexLast.

I'm confused, then. If we randomly decide to request the header at height 1 in the middle of syncing our peer's header chain, we'd do fine in the rest of the ::HEADERS handling here, just (correctly) not requesting any further headers built on top of it. However, we'd immediately disconnect them if we're not in IBD by this new check. Previously this wouldn't break anything, as the next HEADERS they send us will result in us requesting further headers.

sdaftuar replied:
Ah, you're right about that case -- I was thinking of the case where we send a random getheaders after we've already synced the chain, and therefore updated pindexBestKnownBlock to be some recent-ish thing.

I guess it is true that if you just added a random getheaders message in the middle of initial sync, you might not break the sync logic...  I can add a comment indicating that we shouldn't do anything funky until after we've synced a peer's headers chain?

promag commented on src/net_processing.cpp:
```diff
@@ -2383,6 +2401,31 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
             }
         }
+        // If we're in IBD, we want outbound peers that will serve us a useful
+        // chain.  Disconnect peers that are behind us or on chains with
```
Remove extra space.

promag commented on src/net_processing.cpp:
```diff
@@ -3247,6 +3290,52 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+                state.m_sent_getheaders_to_check_chain_sync = false;
+            } else if (state.m_headers_chain_timeout > 0 && time_in_seconds > state.m_headers_chain_timeout) {
+                // No evidence yet that our peer has synced to a chain with work equal to that
+                // of our tip, when we first detected it was behind.  Send a single getheaders
```
Remove extra space.

sdaftuar commented:
Addressed some of @promag's nits at https://github.com/sdaftuar/bitcoin/commits/11490.2, squashed -> 39e0c8e

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -3261,6 +3289,52 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+            // lags behind ours (note: if their chain has more work than ours,
+            // we should sync to it, unless it's invalid, in which case we
+            // should find that out and disconnect from them elsewhere).
+            if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork) {
```
In commit "Permit disconnection of outbound peers on bad/slow chains":

I don't see why the first and second branches of this if/else are distinct. It seems like you could unite them and drop unneeded checks with:

```c++
if (!state.m_header_with_required_work || (state.pindexBestKnownBlock && state.pindexBestKnownBlock->nChainWork >= state.m_header_with_required_work->nChainWork) {
    // Reset timeout and min required work if peer caught up to previous requirement.
    state.m_headers_chain_timeout = time_in_seconds + CHAIN_SYNC_TIMEOUT;
    state.m_header_with_required_work = chainActive.Tip();
    state.m_sent_getheaders_to_check_chain_sync = false;
} else if (time_in_seconds > state.m_headers_chain_timeout) {
...
```

sdaftuar replied:
Hm, I'll think about this.  I haven't quite convinced myself yet that there are no edge-case differences between your simplification and my original patch but you may be right.

ryanofsky commented on src/test/DoS_tests.cpp:
```diff
@@ -42,6 +42,48 @@ static NodeId id = 0;
+    int64_t nStartTime = GetTime();
+    // Wait 21 minutes, to
+    SetMockTime(nStartTime+21*60);
+    peerLogic->SendMessages(&dummyNode1, interruptDummy); // should result in getheaders
```
In commit "Add unit test for outbound peer eviction"

Maybe check dummyNode1.vSendMsg size or contents to see if there actually was a getheaders.

ryanofsky reviewed (commented):
utACK 39e0c8e1e7b5cdf57437aa0f957b41183fd0e197

sdaftuar commented:
@ryanofsky Thanks for the review; I've addressed all but one of your comments, which I'm still thinking about.

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -203,14 +203,32 @@ struct CNodeState {
+    /** State used to enforce CHAIN_SYNC_TIMEOUT
+      * Only in effect for outbound, non-manual connections, with
+      * m_protect == false
+      * Algorithm: if a peer's best known block has less work than our tip,
```
Could maybe be more specific about the timeouts in the description. Instead of "T in the future", could write "CHAIN_SYNC_TIMEOUT seconds in the future", instead of "bump the timeout" could write "reset the timeout", instead of "new shorter timeout" could refer to HEADERS_RESPONSE_TIME.

sdaftuar replied:
Done

ryanofsky commented on src/net_processing.cpp:
```diff
@@ -2383,6 +2383,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+        if (IsInitialBlockDownload() && nCount != MAX_HEADERS_RESULTS) {
+            // When nCount < MAX_HEADERS_RESULTS, we know we have no more
+            // headers to fetch from this peer.
+            if (nodestate->pindexBestKnownBlock->nChainWork < nMinimumChainWork) {
```
Thanks. New "compare their tip" text might make a little more sense as part of the "peer has too little work" comment below rather than the "no more headers to fetch from this peer" comment up here.

sdaftuar commented:
Squashed https://github.com/sdaftuar/bitcoin/commits/11490.3 -> 9dafb45

ryanofsky reviewed (commented):
Thanks for the updates! utACK 9dafb4587baa13ead7fe78ced3deedd303989016

theuni commented on src/net_processing.cpp:
```diff
@@ -2383,6 +2383,24 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+            // won't start block download until we have a headers chain that
+            // has at least nMinimumChainWork, even if a peer has a chain past
+            // our tip, as an anti-DoS measure.
+            if (nodestate->pindexBestKnownBlock->nChainWork < nMinimumChainWork) {
```
I'm unsure if it would be possible to get here without setting a pindexBestKnownBlock, but please double-check to be safe: ```if (nodestate->pindexBestKnownBlock && ...```

sdaftuar replied:
Done

ryanofsky reviewed (commented):
utACK d9a05060ce3c6b8d752898ef68d3dcc0f68c7fba (just adds two suggested comment updates).

theuni commented on src/net_processing.cpp:
```diff
@@ -2396,11 +2436,21 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+        if (!pfrom->fDisconnect && IsOutboundPeer(pfrom)) {
+            // If this is an outbound peer, check to see if we should protect
+            // it from the bad/lagging chain logic.
+            if (g_outbound_peers_with_protect_from_disconnect < MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT && nodestate->pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork && !nodestate->m_chain_sync.m_protect) {
```
```nodestate->pindexBestKnownBlock``` needs a nullptr check here too. Though after a quick look, it doesn't look like we can get here without a pindexBestKnownBlock. If that's the case (I haven't traced it fully) an assert may make more sense.

sdaftuar replied:
Yeah it should be impossible due to the call to UpdateBlockAvailability() above, but I'll add the nullptr check anyway.

theuni commented on src/net_processing.cpp:
```diff
@@ -3265,6 +3315,54 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
 
+        // Check that outbound peers have reasonable chains
+        // GetTime() is used by this anti-DoS logic so we can test this using mocktime
+        int64_t time_in_seconds = GetTime();
```
Can this be factored out into a function? It seems we really only need to call it when we get new headers, or when a timer expires. Moving it out will make it easier to break up the SendMessages monster down the road.

sdaftuar replied:
Done

theuni commented on src/net_processing.cpp:
```diff
@@ -502,6 +532,13 @@ void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vector<con
 
+// Returns true for outbound peers, excluding manual connections, feelers, and
+// one-shots
+bool IsOutboundPeer(const CNode *node)
```
This name is really deceiving. Maybe IsOutboundDisconnectionCandidate?

sdaftuar replied:
done

theuni commented on src/net_processing.cpp:
```diff
@@ -3265,6 +3315,54 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
+                    state.m_chain_sync.m_work_header = nullptr;
+                    state.m_chain_sync.m_sent_getheaders = false;
+                }
+            } else if (state.m_chain_sync.m_timeout == 0 || (state.m_chain_sync.m_work_header != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.m_chain_sync.m_work_header->nChainWork)) {
```
I believe peers can avoid being added as disconnection candidates if they just never respond to requests for blocks/headers (pindexBestKnownBlock is unset). In that case (assuming the "Start block sync" fired off an initial GETHEADERS above), should the clock also start here if pindexBestKnownBlock is null?

Note: This comment conflicts with the one above about possibly asserting pindexBestKnownBlock here. Ignore whichever doesn't apply :)

achow101 replied:
If they don't respond, doesn't the timeout just expire and pindexBestKnownBlock doesn't matter?

sdaftuar replied:
Nice catch, thanks!

theuni replied:
ok, agreed.

sdaftuar commented:
Updated to address latest comments from @theuni.

achow101 commented:
utACK 011423049270dc288bb99d921620f9e0ba0bb43c Needs squash

sdaftuar commented:
Squashed  https://github.com/sdaftuar/bitcoin/commits/11490.4 -> d6c4ad762a1f369a02367bfc18a79e79ab913857

gmaxwell commented:
ACK.  Looks good and I've been running it on a node for a couple days, seen some getheaders but no disconnections yet, though I've started disconnecting healthy outbound peers with the hope of finding some that fail naturally. :)

gmaxwell commented:
After a kicking all my outbounds every couple hours, I eventually hit on this disconnecting a peer... logs show that it was obviously broken. Good job patch.

jnewbery commented:
utACK 453545ffb5b2843757e6644db98f4e8ac11092d2

sdaftuar commented:
@jnewbery pointed out to that the functional test didn't actually need a chain, due to protections given to manual connections (and every connection in regtest is a manual connection).  So I reverted the change to minchainwork and just added a comment that explains the situation.  Also fixed a comment typo that he pointed out in the unit test.

Squashed https://github.com/sdaftuar/bitcoin/commits/11490.6 -> e065249c014a070a8799b2ff947af5b8f012c5c1

jnewbery commented:
utACK e065249c014a070a8799b2ff947af5b8f012c5c1
